crates/stateless/src/validation.rs

```diff
@@ -0,0 +1,292 @@
+/// returns the populated [`SparseStateTrie`] and a [`B256Map`] containing the
+/// contract bytecode (mapping code hash to [`Bytecode`]).
+///
+/// The bytecode has a separate mapping because the [`SparseStateTrie`] does not store the
+/// contract bytecode, only the hash of it (code hash).
+///
+/// If the roots do not match, it returns `None`, indicating the witness is invalid
+/// for the given `pre_state_root`.
+// Note: This approach might be inefficient for ZKVMs requiring minimal memory operations, which
+// would explain why they have for the most part re-implemented this function.
+pub fn verify_execution_witness(
+    witness: &ExecutionWitness,
+    pre_state_root: B256,
+) -> Result<(SparseStateTrie, B256Map<Bytecode>), StatelessValidationError> {
+    let mut trie = SparseStateTrie::new(DefaultBlindedProviderFactory);
+    let mut state_witness = B256Map::default();
+    let mut bytecode = B256Map::default();
+
+    for rlp_encoded in &witness.state {
+        let hash = keccak256(rlp_encoded);
+        state_witness.insert(hash, rlp_encoded.clone());
+    }
+    for rlp_encoded in &witness.codes {
+        let hash = keccak256(rlp_encoded);
+        bytecode.insert(hash, Bytecode::new_raw(rlp_encoded.clone()));
+    }
+
+    // Reveal the witness with our state root
+    trie.reveal_witness(pre_state_root, &state_witness)
+        .map_err(|_e| StatelessValidationError::WitnessRevealFailed { pre_state_root })?;
+
+    // Calculate the root
+    let computed_root = trie
+        .root()
+        .map_err(|_e| StatelessValidationError::StatelessPreStateRootCalculationFailed)?;
+
+    if computed_root == pre_state_root {
+        Ok((trie, bytecode))
+    } else {
+        Err(StatelessValidationError::PreStateRootMismatch {
+            got: computed_root,
+            expected: pre_state_root,
+        })
+    }
+}
+
+/// `BLOCKHASH_HISTORICAL_HASH_LIMIT` specifies the maximum number of historical
+/// block hashes that the [BLOCKHASH](https://www.evm.codes/?fork=cancun#40)
+/// opcode is specified to allow.
+const BLOCKHASH_HISTORICAL_HASH_LIMIT: usize = 256;
+
+/// Verifies the contiguity, number of ancestor headers and extracts their hashes.
+///
+/// This function is used to prepare the data required for the `BLOCKHASH`
+/// opcode in a stateless execution context.
+///
+/// It performs two main checks:
+///
+/// 1. Ensures that the number of provided `ancestor_headers` does not exceed the
+///    `BLOCKHASH_HISTORICAL_HASH_LIMIT` (256). This limit is defined by the `BLOCKHASH` opcode and
+///    has nothing to do with stateless.
+/// 2. Verifies that the provided `ancestor_headers` form a valid, unbroken chain leading back from
+///    the parent of the `current_block`.
+///
+/// Note: This function becomes obsolete if EIP-2935 is implemented.
+///
+/// If both checks pass, it returns a [`HashMap`] mapping the block number of each
+/// ancestor header to its corresponding block hash.
+fn compute_ancestor_hashes(
```

Comment:
add a note on the expected format of `ancestor_headers` (contiguous, sorted in asc order)